File: dart/dartsearch/lib/src/search_options.dart
import 'dart:convert' show json;
import 'dart:io';

import 'package:dartfind/dartfind.dart';
import 'package:dartsearch/src/config.dart' show searchOptionsPath;
import 'package:dartsearch/src/search_settings.dart';

class SearchOption implements Option {
  final String? _shortArg;
  final String _longArg;
  final String _desc;
  final ArgTokenType _argType;

  const SearchOption(this._shortArg, this._longArg, this._desc, this._argType);

  @override
  String? shortArg() => _shortArg;

  @override
  String longArg() => _longArg;

  @override
  String desc() => _desc;

  @override
  ArgTokenType argType() => _argType;

  String sortArg() {
    if (_shortArg == null) {
      return _longArg.toLowerCase();
    } else {
      return '${_shortArg!.toLowerCase()}@${_longArg.toLowerCase()}';
    }
  }

  String optString() {
    if (_shortArg == null) {
      return '--$_longArg';
    } else {
      return '-$_shortArg,--$_longArg';
    }
  }
}

class SearchOptions {
  List<SearchOption> searchOptions = [];
  var boolActionMap = {};
  var stringActionMap = {};
  var intActionMap = {};
  var longArgMap = {'path': 'path'};
  ArgTokenizer? argTokenizer;
  late Future ready;

  SearchOptions() {
    setActionMaps();
    ready = loadSearchOptionsFromJson().then((f) => setArgTokenizer());
  }

  Future<void> loadSearchOptionsFromJson() async {
    var contents = await File(searchOptionsPath).readAsString();
    Map soMap = json.decode(contents);
    if (soMap.containsKey('searchoptions')) {
      var soList = soMap['searchoptions'] as List;
      for (var so in soList) {
        var longArg = (so as Map)['long'];
        var desc = (so)['desc']!;
        String? shortArg;
        if ((so).containsKey('short')) {
          shortArg = (so)['short'];
        }
        var argType = ArgTokenType.unknownType;
        if (boolActionMap.containsKey(longArg)) {
          argType = ArgTokenType.boolType;
        } else if (stringActionMap.containsKey(longArg) ||
            longArg == 'settings-file') {
          argType = ArgTokenType.stringType;
        } else if (intActionMap.containsKey(longArg)) {
          argType = ArgTokenType.intType;
        } else {
          throw FindException('No action for option: $longArg');
        }
        searchOptions.add(SearchOption(shortArg, longArg, desc, argType));
      }
    }
  }

  void setActionMaps() {
    boolActionMap = {
      'archivesonly': (bool b, SearchSettings ss) => ss.archivesOnly = b,
      'allmatches': (bool b, SearchSettings ss) => ss.firstMatch = !b,
      'colorize': (bool b, SearchSettings ss) => ss.colorize = b,
      'debug': (bool b, SearchSettings ss) => ss.debug = b,
      'excludehidden': (bool b, SearchSettings ss) => ss.includeHidden = !b,
      'firstmatch': (bool b, SearchSettings ss) => ss.firstMatch = b,
      'followsymlinks': (bool b, SearchSettings ss) => ss.followSymlinks = b,
      'help': (bool b, SearchSettings ss) => ss.printUsage = b,
      'includehidden': (bool b, SearchSettings ss) => ss.includeHidden = b,
      'multilinesearch': (bool b, SearchSettings ss) => ss.multiLineSearch = b,
      'nocolorize': (bool b, SearchSettings ss) => ss.colorize = !b,
      'nofollowsymlinks': (bool b, SearchSettings ss) => ss.followSymlinks = !b,
      'noprintdirs': (bool b, SearchSettings ss) => ss.printDirs = !b,
      'noprintfiles': (bool b, SearchSettings ss) => ss.printFiles = !b,
      'noprintlines': (bool b, SearchSettings ss) => ss.printLines = !b,
      'noprintmatches': (bool b, SearchSettings ss) => ss.printMatches = !b,
      'noprintresults': (bool b, SearchSettings ss) => ss.printResults = !b,
      'norecursive': (bool b, SearchSettings ss) => ss.recursive = !b,
      'nosearcharchives': (bool b, SearchSettings ss) => ss.searchArchives = !b,
      'printdirs': (bool b, SearchSettings ss) => ss.printDirs = b,
      'printfiles': (bool b, SearchSettings ss) => ss.printFiles = b,
      'printlines': (bool b, SearchSettings ss) => ss.printLines = b,
      'printmatches': (bool b, SearchSettings ss) => ss.printMatches = b,
      'printresults': (bool b, SearchSettings ss) => ss.printResults = b,
      'printusage': (bool b, SearchSettings ss) => ss.printUsage = b,
      'recursive': (bool b, SearchSettings ss) => ss.recursive = b,
      'searcharchives': (bool b, SearchSettings ss) => ss.searchArchives = b,
      'sort-ascending': (bool b, SearchSettings ss) => ss.sortDescending = !b,
      'sort-caseinsensitive': (bool b, SearchSettings ss) =>
          ss.sortCaseInsensitive = b,
      'sort-casesensitive': (bool b, SearchSettings ss) =>
          ss.sortCaseInsensitive = !b,
      'sort-descending': (bool b, SearchSettings ss) => ss.sortDescending = b,
      'uniquelines': (bool b, SearchSettings ss) => ss.uniqueLines = b,
      'verbose': (bool b, SearchSettings ss) => ss.verbose = b,
      'version': (bool b, SearchSettings ss) => ss.printVersion = b,
    };

    stringActionMap = {
      'encoding': (String s, SearchSettings ss) => ss.textFileEncoding = s,
      'in-archiveext': (String s, SearchSettings ss) =>
          ss.addExtensions(s, ss.inArchiveExtensions),
      'in-archivefilepattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.inArchiveFilePatterns),
      'in-dirpattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.inDirPatterns),
      'in-ext': (String s, SearchSettings ss) =>
          ss.addExtensions(s, ss.inExtensions),
      'in-filepattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.inFilePatterns),
      'in-filetype': (String s, SearchSettings ss) =>
          ss.inFileTypes.add(FileTypes.fromName(s)),
      'in-linesafterpattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.inLinesAfterPatterns),
      'in-linesbeforepattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.inLinesBeforePatterns),
      'linesaftertopattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.linesAfterToPatterns),
      'linesafteruntilpattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.linesAfterUntilPatterns),
      'maxlastmod': (String s, SearchSettings ss) =>
          ss.maxLastMod = DateTime.parse(s),
      'minlastmod': (String s, SearchSettings ss) =>
          ss.minLastMod = DateTime.parse(s),
      'out-archiveext': (String s, SearchSettings ss) =>
          ss.addExtensions(s, ss.outArchiveExtensions),
      'out-archivefilepattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.outArchiveFilePatterns),
      'out-dirpattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.outDirPatterns),
      'out-ext': (String s, SearchSettings ss) =>
          ss.addExtensions(s, ss.outExtensions),
      'out-filepattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.outFilePatterns),
      'out-filetype': (String s, SearchSettings ss) =>
          ss.outFileTypes.add(FileTypes.fromName(s)),
      'out-linesafterpattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.outLinesAfterPatterns),
      'out-linesbeforepattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.outLinesBeforePatterns),
      'path': (String s, SearchSettings ss) => ss.paths.add(s),
      'searchpattern': (String s, SearchSettings ss) =>
          ss.addPattern(s, ss.searchPatterns),
      'sort-by': (String s, SearchSettings ss) => ss.sortBy = nameToSortBy(s),
    };

    intActionMap = {
      'linesafter': (int i, SearchSettings ss) => ss.linesAfter = i,
      'linesbefore': (int i, SearchSettings ss) => ss.linesBefore = i,
      'maxdepth': (int i, SearchSettings ss) => ss.maxDepth = i,
      'maxlinelength': (int i, SearchSettings ss) => ss.maxLineLength = i,
      'maxsize': (int i, SearchSettings ss) => ss.maxSize = i,
      'mindepth': (int i, SearchSettings ss) => ss.minDepth = i,
      'minsize': (int i, SearchSettings ss) => ss.minSize = i,
    };
  }

  void setArgTokenizer() {
    argTokenizer = ArgTokenizer(searchOptions);
  }

  Future<void> updateSettingsFromArgTokens(
      SearchSettings settings, List<ArgToken> argTokens) async {
    await ready.then((_) async {
      for (var argToken in argTokens) {
        if (argToken.tokenType == ArgTokenType.boolType) {
          if (boolActionMap.containsKey(argToken.name)) {
            boolActionMap[argToken.name](argToken.value, settings);
          } else {
            throw FindException('Invalid option: ${argToken.name}');
          }
        } else if (argToken.tokenType == ArgTokenType.stringType) {
          if (stringActionMap.containsKey(argToken.name)) {
            stringActionMap[argToken.name](argToken.value, settings);
          } else if (argToken.name == 'settings-file') {
            await updateSettingsFromFile(settings, argToken.value);
          } else {
            throw FindException('Invalid option: ${argToken.name}');
          }
        } else if (argToken.tokenType == ArgTokenType.intType) {
          if (intActionMap.containsKey(argToken.name)) {
            intActionMap[argToken.name](argToken.value, settings);
          } else {
            throw FindException('Invalid option: ${argToken.name}');
          }
        } else {
          throw FindException('Invalid option: ${argToken.name}');
        }
      }
    });
  }

  Future<void> updateSettingsFromJson(
      SearchSettings settings, String jsonString) async {
    await ready.then((_) async {
      List<ArgToken> argTokens = argTokenizer!.tokenizeJson(jsonString);
      await updateSettingsFromArgTokens(settings, argTokens);
    });
  }

  Future<void> updateSettingsFromFile(
      SearchSettings settings, String filePath) async {
    List<ArgToken> argTokens = await argTokenizer!.tokenizeFile(filePath);
    await updateSettingsFromArgTokens(settings, argTokens);
  }

  Future<void> updateSettingsFromArgs(
      SearchSettings settings, List<String> args) async {
    await ready.then((_) async {
      var argTokens = argTokenizer!.tokenizeArgs(args);
      await updateSettingsFromArgTokens(settings, argTokens);
      return settings;
    });
  }

  Future<SearchSettings> settingsFromArgs(List<String> args) async {
    var settings = SearchSettings();
    settings.printResults = true; // default to printing results
    await updateSettingsFromArgs(settings, args);
    return settings;
  }

  Future<String> getUsageString() async {
    return await ready.then((_) {
      var s = 'Usage:\n'
          ' dartsearch [options] -s <searchpattern> <path> [<path> ...]\n\n'
          'Options:\n';
      searchOptions.sort((o1, o2) => o1.sortArg().compareTo(o2.sortArg()));
      var optStrings = searchOptions.map((so) => so.optString()).toList();
      var longest = optStrings.reduce((value, optString) =>
          (optString.length > value.length) ? optString : value);
      for (var i = 0; i < searchOptions.length; i++) {
        s += ' ${optStrings[i].padRight(longest.length + 2, ' ')}';
        s += '${searchOptions[i].desc()}\n';
      }
      return s;
    });
  }

  void usage() async {
    logMsg(await getUsageString());
  }
}

File: fsharp/FsSearch/FsSearchLib/SearchOptions.fs
namespace FsSearchLib

open System
open System.Collections.Generic
open System.Text.Json
open FsFindLib

module SearchOptions =

    type SearchOption = Option

    let boolActionMap : Map<string, bool -> SearchSettings -> Unit> =
        [
            ("allmatches", (fun (b : bool) (settings : SearchSettings) -> settings.FirstMatch <- not b));
            ("archivesonly", (fun (b : bool) (settings : SearchSettings) -> settings.ArchivesOnly <- b));
            ("colorize", (fun (b : bool) (settings : SearchSettings) -> settings.Colorize <- b));
            ("debug", (fun (b : bool) (settings : SearchSettings) -> settings.Debug <- b));
            ("excludehidden", (fun (b : bool) (settings : SearchSettings) -> settings.IncludeHidden <- not b));
            ("firstmatch", (fun (b : bool) (settings : SearchSettings) -> settings.FirstMatch <- b));
            ("followsymlinks", (fun (b : bool) (settings : SearchSettings) -> settings.FollowSymlinks <- b));
            ("help", (fun (b : bool) (settings : SearchSettings) -> settings.PrintUsage <- b));
            ("includehidden", (fun (b : bool) (settings : SearchSettings) -> settings.IncludeHidden <- b));
            ("multilinesearch", (fun (b : bool) (settings : SearchSettings) -> settings.MultiLineSearch <- b));
            ("nocolorize", (fun (b : bool) (settings : SearchSettings) -> settings.Colorize <- not b));
            ("nofollowsymlinks", (fun (b : bool) (settings : SearchSettings) -> settings.FollowSymlinks <- not b));
            ("noprintdirs", (fun (b : bool) (settings : SearchSettings) -> settings.PrintDirs <- not b));
            ("noprintfiles", (fun (b : bool) (settings : SearchSettings) -> settings.PrintFiles <- not b));
            ("noprintlines", (fun (b : bool) (settings : SearchSettings) -> settings.PrintLines <- not b));
            ("noprintmatches", (fun (b : bool) (settings : SearchSettings) -> settings.PrintResults <- not b));
            ("norecursive", (fun (b : bool) (settings : SearchSettings) -> settings.Recursive <- not b));
            ("nosearcharchives", (fun (b : bool) (settings : SearchSettings) -> settings.SearchArchives <- not b));
            ("printdirs", (fun (b : bool) (settings : SearchSettings) -> settings.PrintDirs <- b));
            ("printfiles", (fun (b : bool) (settings : SearchSettings) -> settings.PrintFiles <- b));
            ("printlines", (fun (b : bool) (settings : SearchSettings) -> settings.PrintLines <- b));
            ("printmatches", (fun (b : bool) (settings : SearchSettings) -> settings.PrintResults <- b));
            ("recursive", (fun (b : bool) (settings : SearchSettings) -> settings.Recursive <- b));
            ("searcharchives", (fun (b : bool) (settings : SearchSettings) -> settings.SearchArchives <- b));
            ("sort-ascending", (fun (b : bool) (settings : SearchSettings) -> settings.SortDescending <- not b));
            ("sort-caseinsensitive", (fun (b : bool) (settings : SearchSettings) -> settings.SortCaseInsensitive <- b));
            ("sort-casesensitive", (fun (b : bool) (settings : SearchSettings) -> settings.SortCaseInsensitive <- not b));
            ("sort-descending", (fun (b : bool) (settings : SearchSettings) -> settings.SortDescending <- b));
            ("uniquelines", (fun (b : bool) (settings : SearchSettings) -> settings.UniqueLines <- b));
            ("verbose", (fun (b : bool) (settings : SearchSettings) -> settings.Verbose <- b));
            ("version", (fun (b : bool) (settings : SearchSettings) -> settings.PrintVersion <- b));
        ] |> Map.ofList;

    let stringActionMap : Map<string, string -> SearchSettings -> Unit> =
        [
            ("encoding", (fun (s : string) (settings : SearchSettings) -> settings.TextFileEncoding <- s ))
            ("in-archiveext", (fun (s : string) (settings : SearchSettings) -> settings.InArchiveExtensions <- settings.AddExtensions s settings.InArchiveExtensions))
            ("in-archivefilepattern", (fun (s : string) (settings : SearchSettings) -> settings.InArchiveFilePatterns <- settings.AddPattern s settings.InArchiveFilePatterns))
            ("in-dirpattern", (fun (s : string) (settings : SearchSettings) -> settings.InDirPatterns <- settings.AddPattern s settings.InDirPatterns))
            ("in-ext", (fun (s : string) (settings : SearchSettings) -> settings.InExtensions <- settings.AddExtensions s settings.InExtensions))
            ("in-filepattern", (fun (s : string) (settings : SearchSettings) -> settings.InFilePatterns <- settings.AddPattern s settings.InFilePatterns))
            ("in-filetype", (fun (s : string) (settings : SearchSettings) -> settings.InFileTypes <- settings.AddFileTypes s settings.InFileTypes))
            ("in-linesafterpattern", (fun (s : string) (settings : SearchSettings) -> settings.InLinesAfterPatterns <- settings.AddPattern s settings.InLinesAfterPatterns))
            ("in-linesbeforepattern", (fun (s : string) (settings : SearchSettings) -> settings.InLinesBeforePatterns <- settings.AddPattern s settings.InLinesBeforePatterns))
            ("linesaftertopattern", (fun (s : string) (settings : SearchSettings) -> settings.LinesAfterToPatterns <- settings.AddPattern s settings.LinesAfterToPatterns))
            ("linesafteruntilpattern", (fun (s : string) (settings : SearchSettings) -> settings.LinesAfterUntilPatterns <- settings.AddPattern s settings.LinesAfterUntilPatterns))
            ("maxlastmod", (fun (s : string) (settings : SearchSettings) -> settings.MaxLastMod <- Some(DateTime.Parse(s))));
            ("minlastmod", (fun (s : string) (settings : SearchSettings) -> settings.MinLastMod <- Some(DateTime.Parse(s))));
            ("out-archiveext", (fun (s : string) (settings : SearchSettings) -> settings.OutArchiveExtensions <- settings.AddExtensions s settings.OutArchiveExtensions))
            ("out-archivefilepattern", (fun (s : string) (settings : SearchSettings) -> settings.OutArchiveFilePatterns <- settings.AddPattern s settings.OutArchiveFilePatterns))
            ("out-dirpattern", (fun (s : string) (settings : SearchSettings) -> settings.OutDirPatterns <- settings.AddPattern s settings.OutDirPatterns))
            ("out-ext", (fun (s : string) (settings : SearchSettings) -> settings.OutExtensions <- settings.AddExtensions s settings.OutExtensions))
            ("out-filepattern", (fun (s : string) (settings : SearchSettings) -> settings.OutFilePatterns <- settings.AddPattern s settings.OutFilePatterns))
            ("out-filetype", (fun (s : string) (settings : SearchSettings) -> settings.OutFileTypes <- settings.AddFileTypes s settings.OutFileTypes))
            ("out-linesafterpattern", (fun (s : string) (settings : SearchSettings) -> settings.OutLinesAfterPatterns <- settings.AddPattern s settings.OutLinesAfterPatterns))
            ("out-linesbeforepattern", (fun (s : string) (settings : SearchSettings) -> settings.OutLinesBeforePatterns <- settings.AddPattern s settings.OutLinesBeforePatterns))
            ("path", (fun (s : string) (settings : SearchSettings) -> settings.Paths <- settings.AddPath s settings.Paths))
            ("searchpattern", (fun (s : string) (settings : SearchSettings) -> settings.SearchPatterns <- settings.AddPattern s settings.SearchPatterns))
            ("sort-by", (fun (s : string) (settings : SearchSettings) -> settings.SortBy <- SortUtil.SortByFromName s));
        ] |> Map.ofList

    let intActionMap : Map<string, int -> SearchSettings -> Unit> =
        [
            ("linesafter", (fun (i : int) (settings : SearchSettings) -> settings.LinesAfter <- i))
            ("linesbefore", (fun (i : int) (settings : SearchSettings) -> settings.LinesBefore <- i))
            ("maxdepth", (fun (i : int) (settings : SearchSettings) -> settings.MaxDepth <- i));
            ("maxlinelength", (fun (i : int) (settings : SearchSettings) -> settings.MaxLineLength <- i))
            ("maxsize", (fun (i : int) (settings : SearchSettings) -> settings.MaxSize <- i));
            ("mindepth", (fun (i : int) (settings : SearchSettings) -> settings.MinDepth <- i));
            ("minsize", (fun (i : int) (settings : SearchSettings) -> settings.MinSize <- i));
        ] |> Map.ofList

    type SearchOptionsDictionary = Dictionary<string, List<Dictionary<string,string>>>

    let OptionsFromJson (jsonString : string) : SearchOption list =
        let searchOptionsDict = JsonSerializer.Deserialize<SearchOptionsDictionary>(jsonString)
        let optionDicts = searchOptionsDict["searchoptions"]
        [ for optionDict in optionDicts do
            let longArg = optionDict["long"]
            let shortArg = if optionDict.ContainsKey("short") then optionDict["short"] else ""
            let desc = optionDict["desc"]
            let argType =
                if boolActionMap.ContainsKey(longArg) then ArgTokenType.Bool
                elif stringActionMap.ContainsKey(longArg) then ArgTokenType.Str
                elif intActionMap.ContainsKey(longArg) then ArgTokenType.Int
                else ArgTokenType.Unknown
            yield { ShortArg=shortArg; LongArg=longArg; Description=desc; ArgType=argType } ]

    let _searchOptionsResource = FsSearchLib.EmbeddedResource.GetResourceFileContents("FsSearchLib.Resources.searchoptions.json");
    let options = OptionsFromJson(_searchOptionsResource)

    let GetOptionNameMap : Map<string, string> =
        let shortArgs = seq { for opt in options do if opt.ShortArg <> "" then yield (opt.ShortArg, opt.LongArg) }
        let longArgs =  seq { for opt in options do yield (opt.LongArg, opt.LongArg) }
        Seq.append shortArgs longArgs
        |> Seq.append [("path", "path")]
        |> Map.ofSeq
        
    let optionNameMap = GetOptionNameMap
    
    let argTokenizer : ArgTokenizer =
        ArgTokenizer(options)

    let rec ApplyArgTokenToSettings (argToken : ArgToken) (settings : SearchSettings) : Result<SearchSettings, string> =
        if argToken.Type = ArgTokenType.Bool then
            if boolActionMap.ContainsKey(argToken.Name) then
                match argToken.Value with
                | :? Boolean as b ->
                    boolActionMap[argToken.Name] b settings
                    Ok settings
                | _ -> Error $"Invalid value for option: {argToken.Name}"
            else
                Error $"Invalid value for option: {argToken.Name}"
        elif argToken.Type = ArgTokenType.Str then
            if stringActionMap.ContainsKey(argToken.Name) then
                match argToken.Value with
                | :? string as s ->
                    stringActionMap[argToken.Name] s settings
                    Ok settings
                | _ -> Error $"Invalid value for option: {argToken.Name}"
            else
                Error $"Invalid value for option: {argToken.Name}"
        elif argToken.Type = ArgTokenType.Int then
            if intActionMap.ContainsKey(argToken.Name) then
                match argToken.Value with
                | :? int as i ->
                    intActionMap[argToken.Name] i settings
                    Ok settings
                | _ -> Error $"Invalid value for option: {argToken.Name}"
            else
                Error $"Invalid value for option: {argToken.Name}"
        else
            Error $"Invalid option: {argToken.Name}"

    let UpdateSettingsFromArgTokens (settings : SearchSettings) (argTokens : ArgToken list) : Result<SearchSettings, string> =
        let rec recSettingsFromArgTokens (tokens : ArgToken list) (settings : SearchSettings) : Result<SearchSettings, string> =
            match tokens with
            | [] -> Ok settings
            | token :: tail ->
                match ApplyArgTokenToSettings token settings with
                | Ok settings ->
                    if token.Name = "help" then
                        Ok settings
                    else
                        recSettingsFromArgTokens tail settings
                | Error e -> Error e
        recSettingsFromArgTokens argTokens settings

    let UpdateSettingsFromDictionary (settings : SearchSettings) (dict : Dictionary<string, obj>) : Result<SearchSettings, string> =
        match argTokenizer.TokenizeDictionary(dict) with
        | Ok argTokens -> UpdateSettingsFromArgTokens settings argTokens
        | Error e -> Error e

    let UpdateSettingsFromJson (settings : SearchSettings) (jsonString : string) : Result<SearchSettings, string> =
        match argTokenizer.TokenizeJson(jsonString) with
        | Ok argTokens -> UpdateSettingsFromArgTokens settings argTokens
        | Error e -> Error e

    let SettingsFromJson (jsonString : string) : Result<SearchSettings, string> =
        let settings = SearchSettings()
        UpdateSettingsFromJson settings jsonString

    let UpdateSettingsFromFile (settings : SearchSettings) (filePath : string) : Result<SearchSettings, string> =
        match argTokenizer.TokenizeFile(filePath) with
        | Ok argTokens -> UpdateSettingsFromArgTokens settings argTokens
        | Error e -> Error e

    let SettingsFromFile (filePath : string) : Result<SearchSettings, string> =
        let settings = SearchSettings()
        UpdateSettingsFromFile settings filePath

    let UpdateSettingsFromArgs (settings : SearchSettings) (args : string[]) : Result<SearchSettings, string> =
        match argTokenizer.TokenizeArgs(args) with
        | Ok argTokens -> UpdateSettingsFromArgTokens settings argTokens
        | Error e -> Error e

    let SettingsFromArgs (args : string[]) : Result<SearchSettings, string> =
        let settings = SearchSettings(PrintResults=true)
        UpdateSettingsFromArgs settings args

    let SortOption (o1 : SearchOption) (o2 : SearchOption) : int =
        let os1 = if o1.ShortArg <> "" then o1.ShortArg + "@" + o1.LongArg else o1.LongArg
        let os2 = if o2.ShortArg <> "" then o2.ShortArg + "@" + o2.LongArg else o2.LongArg
        String.Compare(os1, os2, StringComparison.OrdinalIgnoreCase)

    let GetUsageString () : string =
        let sortedOptions = options |> List.sortWith SortOption
        let optStringMap =
            [ for opt in sortedOptions do
                let shortString : string = 
                    if opt.ShortArg <> "" then "-" + opt.ShortArg + ","
                    else ""
                let longString : string = "--" + opt.LongArg
                yield (opt.LongArg, shortString + longString) ]
            |> Map.ofList

        let optDescMap =
            [ for opt in options do
                yield (opt.LongArg, opt.Description) ]
            |> Map.ofList

        let optStrings : string list =
            optStringMap
            |> Map.toList
            |> List.map snd

        let longest = 
            optStrings
            |> Seq.map (fun o -> o.Length)
            |> Seq.max

        let format = " {0,-" + longest.ToString() + "}  {1}"

        let usageStrings =
            [for o in options do
                yield String.Format(format, optStringMap[o.LongArg], optDescMap[o.LongArg])]

        let usageString = 
            usageStrings
            |> List.append ["\nUsage:"; " fssearch [options] -s <searchpattern> <path> [<path> ...]\n"; "Options:"] 
            |> String.concat "\n"
        usageString

    let Usage (exitCode : int) : unit =
        let usageString = GetUsageString()
        printfn $"%s{usageString}\n"
        Environment.Exit(exitCode)
